File: app/src/main/java/com/github/windsekirun/yukarisynthesizer/di/AppProvidesModule.java
package com.github.windsekirun.yukarisynthesizer.di;

import com.github.windsekirun.yukarisynthesizer.MainApplication;
import com.github.windsekirun.yukarisynthesizer.net.JSONService;
import com.github.windsekirun.yukarisynthesizer.repository.MainRepository;
import com.github.windsekirun.yukarisynthesizer.repository.PreferenceRepository;
import com.github.windsekirun.yukarisynthesizer.repository.PreferenceRepositoryImpl;
import dagger.Module;
import dagger.Provides;
import pyxis.uzuki.live.richutilskt.utils.RPreference;
import retrofit2.Retrofit;

import javax.inject.Singleton;

@Module
public class AppProvidesModule {

    @Singleton
    @Provides
    JSONService provideJSONService(Retrofit retrofit) {
        return retrofit.create(JSONService.class);
    }

    @Provides
    @Singleton
    public RPreference provideRPerference(MainApplication application) {
        return RPreference.getInstance(application);
    }

    @Provides
    @Singleton
    PreferenceRepository providePreferenceRepository(PreferenceRepositoryImpl impl) {
        return impl;
    }

    @Provides
    @Singleton
    PreferenceRepositoryImpl providePreferenceRepositoryImpl(MainApplication application) {
        return new PreferenceRepositoryImpl(application);
    }

    @Provides
    @Singleton
    MainRepository provideMainRepository(JSONService service) {
        return new MainRepository(service);
    }
}

File: app/src/main/java/com/github/windsekirun/yukarisynthesizer/di/AppProvidesModule.kt
package com.github.windsekirun.yukarisynthesizer.di

import com.github.windsekirun.yukarisynthesizer.MainApplication
import com.github.windsekirun.yukarisynthesizer.core.YukariOperator
import com.github.windsekirun.yukarisynthesizer.core.repository.PreferenceRepository
import com.github.windsekirun.yukarisynthesizer.core.repository.PreferenceRepositoryImpl
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/**
 * Dagger module for provide project-specific class
 */
@Module
class AppProvidesModule {

    @Provides
    @Singleton
    fun providePreferenceRepository(impl: PreferenceRepositoryImpl): PreferenceRepository {
        return impl
    }

    @Provides
    @Singleton
    fun providePreferenceRepositoryImpl(context: MainApplication): PreferenceRepositoryImpl {
        return PreferenceRepositoryImpl(context)
    }

    @Provides
    @Singleton
    fun provideYukariOperator(context: MainApplication): YukariOperator {
        return YukariOperator(context)
    }
}